static/js/import_wizard/utils/detection.ts

```diff
@@ -0,0 +1,216 @@
+  Mapping,
+  MappingType,
+  MappingVal,
+  TypeProperty,
+} from "../types";
+
+const COLUMN_API_KEY = "column";
+const COLUMN_ID_API_KEY = "id";
+const COLUMN_HEADER_API_KEY = "header";
+const COLUMN_IDX_API_KEY = "column_idx";
+const DCID_API_KEY = "dcid";
+const DISP_NAME_API_KEY = "display_name";
+const PLACETYPE_API_KEY = "place_type";
+const PLACEPROP_API_KEY = "place_property";
+const DCTYPE_API_KEY = "dc_type";
+const DCPROP_API_KEY = "dc_property";
+const HEADER_API_KEY = "headers";
+
+function isDCTypeOrProperty(entity): boolean {
+  if (_.has(entity, DCID_API_KEY) && _.has(entity, DISP_NAME_API_KEY)) {
+    return true;
+  }
+  return false;
+}
+
+function isColumn(candidate): boolean {
+  if (
+    _.has(candidate, COLUMN_ID_API_KEY) &&
+    _.has(candidate, COLUMN_HEADER_API_KEY) &&
+    _.has(candidate, COLUMN_IDX_API_KEY)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Given the API response from the detection server API,
+ * parse it in to a Mapping structure.
+ *
+ * @param detectedResponse is the Object response from the server API.
+ *
+ * @returns a Mapping structure which has details of all the detected
+ *  columns. If there are any unexpected validation errors, that particular
+ *  MappingVal is skipped.
+ */
+export function parseDetectionAPIResponse(detectedResponse): Mapping {
+  const mToReturn: Mapping = new Map<MappedThing, MappingVal>();
+
+  // Some validations.
+  for (const [key, mVal] of Object.entries(detectedResponse)) {
+    // Keys must be among the MappedThing enum.
+    if (!_.includes(Object.values(MappedThing), key)) {
+      continue;
+    }
+    // Key "type" must be present and the value must be among MappingType enum.
+    if (
+      !_.includes(Object.keys(mVal), "type") ||
+      !_.includes(Object.values(MappingType), mVal["type"])
+    ) {
+      continue;
+    }
+    const mValValidated: MappingVal = { type: mVal["type"] };
+
+    // If key "column" is present then the value must be of type Column.
+    let col: Column;
+    if (_.includes(Object.keys(mVal), COLUMN_API_KEY) && mVal[COLUMN_API_KEY]) {
+      const candidateCol = mVal[COLUMN_API_KEY];
+      if (isColumn(candidateCol)) {
+        col = {
+          id: candidateCol[COLUMN_ID_API_KEY],
+          header: candidateCol[COLUMN_HEADER_API_KEY],
+          columnIdx: candidateCol[COLUMN_IDX_API_KEY],
+        };
+        mValValidated.column = col;
+      }
+    }
+
+    // If key "placeType" is present then the value must be of type DCType.
+    // Note, if the validation succeeds, the value (DCType) is replaced by a map of
+    // columnIdx (number) to the DCType.
+    if (
+      _.includes(Object.keys(mVal), PLACETYPE_API_KEY) &&
+      mVal[PLACETYPE_API_KEY]
+    ) {
+      if (!isDCTypeOrProperty(mVal[PLACETYPE_API_KEY])) {
+        continue;
+      }
+      // If made this far, then mVal must also have had the column field. Extract the columnIdx
+      // and assign it as a key to mVal["placeProperty"].
```

Comment:
should it be an error if there is no column, instead of a silent continue?

Reply:
How are we logging client side errors? If you point me to it, I can follow the convention here. Sorry for overlooking

Reply:
In other apps we display an error message. I'm not sure if that's appropriate here.. let's add a todo for now?

Reply:
Added a todo at the top